File: sdk/js/lima-sdk/src/resources/devices.js
import { raiseForStatus } from "../client.js";

/** @typedef {import('../client.js').LiMaClient} LiMaClient */

export class Devices {
  /**
   * @param {LiMaClient} client
   */
  constructor(client) {
    this.client = client;
  }

  /**
   * @returns {Promise<{devices: Record<string, unknown>[], count: number}>}
   */
  async list() {
    return this._getJson("/device/v1/app/devices");
  }

  /**
   * @param {string} deviceId
   * @returns {Promise<Record<string, unknown>>}
   */
  async get(deviceId) {
    return this._getJson(`/device/v1/app/devices/${deviceId}`);
  }

  /**
   * @param {string} deviceId
   * @returns {Promise<Record<string, unknown>>}
   */
  async status(deviceId) {
    return this._getJson(`/device/v1/app/devices/${deviceId}/status`);
  }

  /**
   * @param {string} deviceId
   * @param {Record<string, unknown>} body
   * @returns {Promise<Record<string, unknown>>}
   */
  async createTask(deviceId, body) {
    return this._postJson(`/device/v1/app/devices/${deviceId}/tasks`, body);
  }

  /**
   * @param {string} deviceId
   * @param {Record<string, string|number>} [params]
   * @returns {Promise<{tasks: Record<string, unknown>[], count: number}>}
   */
  async listTasks(deviceId, params = {}) {
    const query = new URLSearchParams({ device_id: deviceId, ...stringifyParams(params) });
    return this._getJson(`/device/v1/app/tasks?${query.toString()}`);
  }

  /**
   * @param {string} taskId
   * @returns {Promise<Record<string, unknown>>}
   */
  async getTask(taskId) {
    return this._getJson(`/device/v1/app/tasks/${taskId}`);
  }

  /**
   * @param {string} path
   */
  async _getJson(path) {
    const response = await this.client.request("GET", path);
    await raiseForStatus(response);
    return response.json();
  }

  /**
   * @param {string} path
   * @param {Record<string, unknown>} body
   */
  async _postJson(path, body) {
    const response = await this.client.request("POST", path, { json: body });
    await raiseForStatus(response);
    return response.json();
  }
}

/**
 * @param {Record<string, string|number>} params
 * @returns {Record<string, string>}
 */
function stringifyParams(params) {
  /** @type {Record<string, string>} */
  const out = {};
  for (const [key, value] of Object.entries(params)) {
    out[key] = String(value);
  }
  return out;
}

File: sdk/go/lima/devices.go
package lima

import (
	"context"
	"fmt"
)

// Device represents a device returned by the API.
type Device map[string]any

// Task represents a device task returned by the API.
type Task map[string]any

// DevicesService manages devices and tasks.
type DevicesService struct {
	client *Client
}

// List returns all devices bound to the account.
func (s *DevicesService) List(ctx context.Context) ([]Device, error) {
	var result struct {
		Devices []Device `json:"devices"`
	}
	if err := s.client.jsonRequest(ctx, "GET", "/device/v1/app/devices", nil, &result); err != nil {
		return nil, err
	}
	return result.Devices, nil
}

// Get returns a single device by ID.
func (s *DevicesService) Get(ctx context.Context, deviceID string) (Device, error) {
	var dev Device
	if err := s.client.jsonRequest(ctx, "GET", fmt.Sprintf("/device/v1/app/devices/%s", deviceID), nil, &dev); err != nil {
		return nil, err
	}
	return dev, nil
}

// Status returns the runtime status of a device.
func (s *DevicesService) Status(ctx context.Context, deviceID string) (map[string]any, error) {
	var status map[string]any
	if err := s.client.jsonRequest(ctx, "GET", fmt.Sprintf("/device/v1/app/devices/%s/status", deviceID), nil, &status); err != nil {
		return nil, err
	}
	return status, nil
}

// CreateTask creates a task on a device.
func (s *DevicesService) CreateTask(ctx context.Context, deviceID string, body map[string]any) (Task, error) {
	var task Task
	path := fmt.Sprintf("/device/v1/app/devices/%s/tasks", deviceID)
	if err := s.client.jsonRequest(ctx, "POST", path, body, &task); err != nil {
		return nil, err
	}
	return task, nil
}

// ListTasks returns tasks for a device.
func (s *DevicesService) ListTasks(ctx context.Context, deviceID string, params map[string]string) ([]Task, error) {
	path := "/device/v1/app/tasks?device_id=" + deviceID
	for k, v := range params {
		path += "&" + k + "=" + v
	}
	var result struct {
		Tasks []Task `json:"tasks"`
	}
	if err := s.client.jsonRequest(ctx, "GET", path, nil, &result); err != nil {
		return nil, err
	}
	return result.Tasks, nil
}

// GetTask returns a single task by ID.
func (s *DevicesService) GetTask(ctx context.Context, taskID string) (Task, error) {
	var task Task
	if err := s.client.jsonRequest(ctx, "GET", fmt.Sprintf("/device/v1/app/tasks/%s", taskID), nil, &task); err != nil {
		return nil, err
	}
	return task, nil
}
